src/credentials/FabricTable.cpp

```diff
@@ -54,6 +54,10 @@ constexpr TLV::Tag kOpKeyDataTag    = TLV::ContextTag(1);
 // If this version grows beyond UINT16_MAX, adjust OpKeypairTLVMaxSize
 // accordingly.
 constexpr uint16_t kOpKeyVersion = 1;
+
+// Tags for our index list storage.
+constexpr TLV::Tag kNextAvailableFabricIndexTag = TLV::ContextTag(0);
+constexpr TLV::Tag kFabricIndicesTag            = TLV::ContextTag(1);
```

Comment:
Can you outline the implications this has for atomicity in our fabric table storage?  It makes me very nervous for the sdk to seemingly have a reliance on storing of up to three kv entries without interruption in order to avoid corrupting our kvstore.

I know that some platforms will always be at risk of persisting inconsistent data due to fundamental, underlying limitations.  But for platforms that can have transactional writes, it would be great if either a) each fabric table index was fully self-describing or b) we had a kvstore 'commit' method.

Reply:
Right now adding a fabric requires storing more like 4-5 keys (NOC, ICAC, etc, etc, plus after this PR the one key for the updated fabric index state).  That's for the fabric table.  It also requires writes for group keys (IPK), ACLs, and maybe other things.

Removing a fabric requires multiple remove calls (NOC, ICAC, etc) and after this PR one store for the fabric index state.  Plus removals of the other state for the fabric.

We could make fabric table indices somewhat more self-describing if we had a way to enumerate keys, but even then the "what's the last index we assigned" would be separate state....  And of course each entry would still be multiple keys.

And yes, if we could do transactional writes, that would be pretty nice.  That would be useful in lots of places, not just the fabric table...

Reply:
Thanks for describing the current state.

I think then my question is whether this PR is adding additional fragility by introducing kvstore entries that, if their write is missed, can corrupt all fabrics.  If so, are there any possible mitigations?

Reply:
```
            err = Store(newFabricIndex);
            if (err != CHIP_NO_ERROR)
            {    
                fabric.Reset();
                FabricInfo::DeleteFromStorage(mStorage, newFabricIndex);
                return err; 
            }    

            UpdateNextAvailableFabricIndex();
            err = StoreFabricIndexInfo();
```

So looking at this then, I think my only actionable critique is that storing fabric index info after storing the associated fabrics means we can effectively lose fabrics on disk.  Fabrics are large relative to some platforms' storage.  We don't want to leak them.

This should be fixable by simply reversing the calls to `Store(newFabricIndex);` and `StoreFabricIndexInfo();`

The other alternative (and one I'm assuming you're trying to avoid) is to always try all 255 possible fabric indices so that we don't have to store fabric indices separately.  If our only priority was on-disk consistency, that would be a step in that direction.

Reply:
Ah, and now I see this still doesn't operate entirely as I was hoping because the fabric info isn't stored atomically (RCAC, ICAC, NOC, are separate as you pointed out).

I don't know what the solution is.  But this PR means we can now leak on-disk fabric information across all 255 possible fabric indices instead of CHIP_CONFIG_MAX_FABRICS.  That seems notable.

Reply:
> I think then my question is whether this PR is adding additional fragility by introducing kvstore entries that, if their write is missed, can corrupt all fabrics.

That is a good question.  I am assuming that by "write is missed" you mean device crashes or shuts down in some other way right before we try to write this new key, right?  That is, we are not worrying about partial writes of the single key/value pair involved, where part of the value is written out but not all of it)?

If so, there are two places we write this new key: In `AddNewFabric` and in `Delete`.

* `AddNewFabric`: The write to the new key happens after all the other FabricTable-related storage operations here.  If the write never happens then when we later boot up we will not realize the fabric has been added.  Which is fine, because we also never replied with a success response to the caller of AddNOC, so they know the fabric has not been added, once they time out waiting for their response.

    Because we can't associate the fail-safe context with the new fabric index until `AddNewFabric` returns (since we don't know what the fabric index for the new fabric is), fail-safe cleanup on boot will not remove the state that `AddNewFabric` had stored before the shutdown, which might include the NOC for the fabric, etc, etc.  But whenever we try to add a new fabric again and end up reusing that fabric index, that state will be overwritten by the state for the new fabric.  So I don't think we ever end up in a corrupt state here.

    Arguably this PR actually makes this codepath _better_.  Before this PR, we could store part of the state in `AddNewFabric`, then get shut down, and then on restart we would find that partial state in storage and read it out into our fabric table, ending up in a weird state.  After this PR, since the last thing we do is write out the set of fabric indices, this can't happen.

* `Delete`: Here we remove the fabric's data from storage first and then store the updated fabric index info.  So the failure mode here would be that we have a fabric index in our stored list that no longer has any data backing it, or where at least some of the data backing it is gone.

    When we boot we will go through the indices in our list and try to load the fabric info for them all.  When we get to the index where some of the data is gone, we will fail the LoadFromStorage call, and effectively skip over that fabric index.  Our resulting fabric table will not contain a fabric at that index.  The next time a fabric is added or removed we will update the fabric-index-list key's value to match reality.

    This is pretty much the state of things for `Delete` before this PR too, as far as I can tell: We would just try to read for all the indices up to MAX_FABRICS and whichever ones failed, well, we have no fabric with that index.

Reply:
> That is a good question. I am assuming that by "write is missed" you mean device crashes or shuts down in some other way right before we try to write this new key, right? That is, we are not worrying about partial writes of the single key/value pair involved, where part of the value is written out but not all of it)?

Yes, exactly.  The most our existing kvstore interface can provide is atomic writes of individual entries.  I know the API doesn't promise this, but it's possible for a platform to provide it.  So given that as our boundary for best possible behavior, I was trying to evaluate cases where shutdown or reset in between writes could put a node into an unrecoverable state, and whether this PR was increasing the number of ways this can happen.

Based on your detailed description, the behavior sounds pretty reasonable.